orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/support/TargetServerGroupLinearStageSupport.groovy

```diff
@@ -16,77 +16,173 @@
 
+/**
+ * Stages extending this class will resolve targets then generate synthetic before
+ * stages (of the same type) to operate on each target.
+ */
 @Slf4j
 abstract class TargetServerGroupLinearStageSupport implements StageDefinitionBuilder, Nameable {
 
-  @Autowired
-  TargetServerGroupResolver resolver
+  @Autowired TargetServerGroupResolver resolver
 
-  @Autowired
-  DetermineTargetServerGroupStage determineTargetServerGroupStage
+  /**
+   * Override to supply tasks that individual target stages will run. The top level
+   * stage will never run any tasks.
+   */
+  protected void taskGraphInternal(Stage stage, TaskNode.Builder builder) {}
 
-  String name = this.type
+  /**
+   * Override to supply before stages for individual target stages operating on a
+   * static target.
+   */
+  protected void preStatic(Map<String, Object> descriptor, StageGraphBuilder graph) {
+  }
+
+  /**
+   * Override to supply after stages for individual target stages operating on a
+   * static target.
+   */
+  protected void postStatic(Map<String, Object> descriptor, StageGraphBuilder graph) {
+  }
+
+  /**
+   * Override to supply before stages for individual target stages operating on a
+   * dynamic target.
+   */
+  protected void preDynamic(Map<String, Object> context, StageGraphBuilder graph) {
+  }
+
+  /**
+   * Override to supply after stages for individual target stages operating on a
+   * dynamic target.
+   */
+  protected void postDynamic(Map<String, Object> context, StageGraphBuilder graph) {
+  }
 
   @Override
-  def List<Stage> aroundStages(Stage parentStage) {
-    return composeTargets(parentStage)
+  String getName() {
+    return type
   }
 
-  @Memoized
-  List<Stage> composeTargets(Stage stage) {
-    stage.resolveStrategyParams()
-    def params = TargetServerGroup.Params.fromStage(stage)
-    if (TargetServerGroup.isDynamicallyBound(stage)) {
-      return composeDynamicTargets(stage, params)
+  @Override
+  final void taskGraph(Stage stage, TaskNode.Builder builder) {
+    if (!isTopLevel(stage)) {
+      // Tasks are only run by individual target stages
+      taskGraphInternal(stage, builder)
     }
+  }
 
-    return composeStaticTargets(stage, params)
+  @Override
+  final void beforeStages(
+    @Nonnull Stage parent,
+    @Nonnull StageGraphBuilder graph
+  ) {
+    if (isTopLevel(parent)) {
+      // the top level stage should resolve targets and create synthetic stages to
+      // deal with each one
+      composeTargets(parent, graph)
+    } else {
+      // a non top level stage operates on a single target and may have its own
+      // synthetic stages
+      if (isDynamicallyBound(parent)) {
+        preDynamic(parent.context, graph)
+      } else {
+        preStatic(parent.context, graph)
+      }
+    }
   }
 
-  private List<Stage> composeStaticTargets(Stage stage, TargetServerGroup.Params params) {
-    if (stage.parentStageId) {
```

Comment:
I've replicated this logic with my `isTopLevel` method but I'm concerned it's not quite correct. Shouldn't the check be whether the parent is the same type as this stage?